### ISSUE TITLE
Support for injection into managed/.NET processes?

### ISSUE THREAD

**Issue by lahma0** [Mbucari/InjectDotnet#1]
Sorry, I know filing an "Issue" to ask a simple question such as this isn't really proper but I was not sure how else to ask a question. Does this injector support injection into managed processes (.NET Framework or .NET/.NET Core)? For example, injecting a .NET DLL which uses Harmony to patch a running managed process?

Looks like you spent a lot of time and effort on writing this entire framework. Great work!

**Comment by Mbucari** [comment]
> Sorry, I know filing an "Issue" to ask a simple question such as this isn't really proper but I was not sure how else to ask a question. 

Not a problem. I should probably enable discussions on this repo.

> Does this injector support injection into managed processes (.NET Framework or .NET/.NET Core)? For example, injecting a .NET DLL which uses Harmony to patch a running managed process?

You _can_ inject into a manages process, but there are some limitations that I don't fully understand.  Firstly, InjectDotNet currently only supports injecting runtime-dependent Dlls. You can't (yet) use it to inject into a self-contained target. Also, it appears that the injected Dll's runtime version must be less than or equal to the target's runtime version. Finally, you can't (yet) inject a self-contained dll.

That said, if you satisfy those requirements then the injected Dll can be injected and will have access to the targets AppDomain (since .NET Core only allows one AppDomain per process).

> Looks like you spent a lot of time and effort on writing this entire framework. Great work!

Thank you! It was quite a journey, but I'm quite proud of it.

**Comment by Mbucari** [comment]
@lahma0 I made a [new release](https://github.com/Mbucari/InjectDotnet/releases/latest) that improves injection into .NET managed process and added [documentation for some of the limitations](https://github.com/Mbucari/InjectDotnet/tree/master#injecting-into-net-targets).

**Comment by Mbucari** [comment]
Regarding .NET Framework, I don't think InjectDotNet will be useful for your purpose. It only injects .NET Core Dlls, not .NET Framework. You _can_ inject a .NET Core Dll into a .NET Framework proces, but the injected Dll won't be able to communicate with it through reflection.

**Comment by Mbucari** [comment]
@lahma0 Let me ask you, do you think there's much use for a .NET Framework Dll injector these days? I could re-open this issue and make it a TODO if it would be useful.

**Comment by lahma0** [comment]
> @lahma0 Let me ask you, do you think there's much use for a .NET Framework Dll injector these days? I could re-open this issue and make it a TODO if it would be useful.

Hey @Mbucari, thanks for replying to my message! Honestly, I'm not sure how much demand there would be for a .NET Framework injector these days.. Obviously there are still a LOT of popular apps that utilize .NET Framework, which never updated to .NET Core (or rather just ".NET" now), but I suspect most newly built apps, or projects still in active development (not just maintenance releases), are using .NET/.NET Core. 

Just as an example, I was most recently looking for a .NET Framework injector for the purposes of dynamically patching some methods in a popular screen clipping/screenshot app called ShareX. The 2 injectors I found that seemed to work best were Holly Hacker's Managed Injector (only supports injecting .NET Framework libraries into .NET Framework processes) and Extreme Injector (supports injecting native DLLs as well as .NET Core and .NET Framework with the help of an additional library). With Extreme Injector, in order to inject managed libraries into managed processes, your injected DLL must make use of a library which allows you to declare export functions in the export address table. Obviously, this isn't something that is natively supported by .NET assembles so you must make use of a library such as DNNE ( https://github.com/AaronRobinsonMSFT/DNNE ) to create an export function which Extreme Injector will then recognize and allow you to select before injecting the DLL. The only issue I had with Extreme Injector was that it didn't appear to support command line injection; everything had to be done through the GUI. That's fine for one off uses but if you have a DLL you want to inject automatically every time an app starts (or when your computer boots), you don't want to have to do that manually through a GUI every single time. 

If you were to implement the ability to inject .NET Framework DLLs into .NET Framework processes, I know I would certainly use it. It would be really nice to have a single injector that supports injecting .NET/.NET Core assemblies into .NET/.NET Core processes, .NET Framework assemblies into .NET Framework processes, AND .NET/.NET Core/.NET Framework assemblies into native processes. As far as I'm aware, there isn't a single injector that supports all of the above. I guess you could argue that Extreme Injector kind of does but that's not really the case since you must use 3rd party libraries in your injected assembly (to expose an export function to native code). 

Harmony has seemingly become the go-to, one stop shop, for runtime patching of managed assemblies regardless of their type (any version of .NET Framework or .NET/.NET Core). It is extremely adept at everything from basic patches of static methods to pre/post patches of property getters/setters to advanced replacement of generic methods. Unfortunately, it doesn't have an injector to go with it though. It would be nice to have a one-stop-shop injector to pair with it for the ultimate injector+patcher combination.

**Comment by Mbucari** [comment]
@lahma0 I've managed to get this working with .NET Framework and can inject dlls into a target process's Default AppDomain. I haven't figured out a way to inject into an arbitrary AppDomain, but AppDomains are such a rarely-used feature that this is hardly a limitation.

~~One thing I'm not prepared to do is downgrade InjectDotnet.NativeHelper to netstandard2.0. It makes heavy use of C# 8.0+ features, and it would be a massive PITA to do.~~

Well, I changed my tune on that pretty quickly. I'm still working through a couple of bugs and quality control issues, but I will package this for netstandard2 and net6. The biggest problem is not having `nint`, but I'm working around that.

**Comment by Mbucari** [comment]
@lahma0 [v0.4.0](https://github.com/Mbucari/InjectDotnet/releases/tag/0.4.0) adds support for .NET Framework. Please try it out and let me know if there are any issues.

**Comment by lahma0** [comment]
> @lahma0 [v0.4.0](https://github.com/Mbucari/InjectDotnet/releases/tag/0.4.0) adds support for .NET Framework. Please try it out and let me know if there are any issues.

That is great news @Mbucari! To me, it's not really a big deal not being able to inject into an arbitrary AppDomain on .NET Framework. Like you stated, it is pretty rare to come across apps that actually make use of multiple AppDomains so its unlikely to be a relevant limitation for most people. I will test out the new build asap and report back here with the results. Are you ok with me continuing to reply in this thread now that it is closed?

Perhaps I'm just not understanding the documentation but I noticed in your README.md that you mention that injecting into a new managed process at startup is not supported: https://github.com/Mbucari/InjectDotnet/blob/9d854372594b85432edb9b4c8058a239bff5d9a4/README.md?plain=1#L43 However, immediately following that, is a section that describes how to inject into a new process at startup: https://github.com/Mbucari/InjectDotnet/blob/9d854372594b85432edb9b4c8058a239bff5d9a4/README.md?plain=1#L47-L51 Now that I'm actually thinking about it though, perhaps that is explaining how to inject a managed DLL into an unmanaged process at startup and injecting a managed DLL into a managed process at startup is indeed unsupported. That would make sense (though a little more clarity might be helpful).

**Comment by Mbucari** [comment]
> Now that I'm actually thinking about it though, perhaps that is explaining how to inject a managed DLL into an unmanaged process at startup and injecting a managed DLL into a managed process at startup is indeed unsupported. That would make sense (though a little more clarity might be helpful).

That is exactly correct, and I will clarify the documentation.

**Comment by Mbucari** [comment]
@lahma0 

> Are you ok with me continuing to reply in this thread now that it is closed?

For general discussion about injecting into dotnet framework sure, but if you find specific bugs then I would appreciate opening a new issue to make tracking easier.

**Comment by lahma0** [comment]
So I finally got everything set up to test out using your injector to inject my DLL into a .NET Framework process. I'm having an issue though and I don't really know how to debug/troubleshoot it. Essentially, everything appears to be working correctly but ultimately the bootstrap function is never executed. The DLL IS injected into the process and the "Inject" method is not returning an error code (I do have waitForReturn set to true) but my breakpoint inside the bootstrap function is never being triggered (and the patches which are supposed to be made to the assembly by Harmony definitely aren't being made). I know this DLL works for patching the assembly because like I explained previously, I've used it with 2 different injectors and it is working fine.

The bootstrap method's signature looks like this:
`public static int MyMethod(string arg)`

I'm using dnSpy to set breakpoints on the bootstrap method, attaching to the process with dnSpy, and then running the injector. Like I said, I've inspected the running process and the DLL is definitely being injected into the process (and my breakpoints get enabled in dnSpy as soon as the injector is run) but the bootstrap method just isn't executed. Any tips on how to further debug/troubleshoot the issue?

Just for your reference, here is my injector's code:
```
using System.Diagnostics;
using System.Reflection;
using System.Runtime.InteropServices;
using InjectDotnet;

namespace ShareXInjector
{
    internal class Program
    {
        private const string ProcessName = "ShareX";
        private const string InjectDllName = "ShareXPatcher.dll";

        private static readonly string InjectDllPath = Environment.ExpandEnvironmentVariables(
            $"%PortableAppsDir%\\graphics-image_editing-design\\ShareX\\{InjectDllName}");

        private const string InjectDllNamespace = "ShareXPatcher.Class";
        private const string InjectDllMethod = "MyMethod";

        private static readonly string? InjectDllTypeInfo = Assembly.LoadFile(InjectDllPath)?.GetType(InjectDllNamespace)?.AssemblyQualifiedName;

        private const bool WaitForReturn = true;
        static void Main(string[] args)
        {

            if (InjectDllTypeInfo == null)
            {
                Console.WriteLine($"Cannot inject DLL because '{nameof(InjectDllTypeInfo)}' is null. This is likely a result of not being able to read the file at '{InjectDllPath}'.");
                return;
            }

            var targetProcess = Process.GetProcessesByName(ProcessName).FirstOrDefault();
            if (targetProcess == null)
            {
                Console.WriteLine($"Cannot inject DLL because '{ProcessName}' is not running.");
                return;
            }

            Console.WriteLine($"Injecting '{InjectDllPath}' into {ProcessName} process ({targetProcess.Id})...");
            try
            {
                var errorCode = targetProcess.Inject(InjectDllPath, InjectDllTypeInfo, InjectDllMethod, string.Empty, WaitForReturn);
                if (!WaitForReturn)
                {
                    Console.WriteLine("Injection complete!");
                    return;
                }

                if (errorCode > 0)
                {
                    Marshal.ThrowExceptionForHR((int)errorCode);
                }
                else
                {
                    Console.WriteLine("Injection completed successfully!");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Injection failed: {ex}");
            }
            
        }
    }
}
```
And the injector's console output is:
```
.\ShareXInjector.exe
Injecting 'R:\PortableApps\graphics-image_editing-design\ShareX\ShareXPatcher.dll' into ShareX process (61300)...
Injection completed successfully!
```
Oh, and just in case it is useful, here's my boostrap method:
```
public static int MyMethod(string arg)
{
    Trace.WriteLine($"'{nameof(ShareXPatcher)}' injected successfully. Now patching code using Harmony " + 
                    $"instance ID '{HarmonyInstanceId}'...");
    _harmony = new Harmony(HarmonyInstanceId);
    _harmony.PatchAll();
    Trace.WriteLine("Patches complete.");
    return 0;
}
```

**Comment by Mbucari** [comment]
@lahma0 Sorry for the confusion, I guess I need to clarify the documentation.

`AssemblyQualifiedName` is used when injecting .NET Core Dlls. .NET Framework Dlls just use the type name. So replace 
```
var errorCode = targetProcess.Inject(InjectDllPath, InjectDllTypeInfo, InjectDllMethod, string.Empty, WaitForReturn);
```

with this

```
var errorCode = targetProcess.Inject(InjectDllPath, InjectDllNamespace, InjectDllMethod, string.Empty, WaitForReturn);

```
I've modified your code and tested that it works with ShareX
[ShareXInjector.zip](https://github.com/Mbucari/InjectDotnet/files/13694914/ShareXInjector.zip)


EDIT* I think I'm going to add overload Inject methods that accept `MethodInfo` instead of string type name and method name.

**Comment by Mbucari** [comment]
I'm actually surprised your code didn't return an error code. I ran your injector verbatim and it returned `0x80131522`, which is a [COR_E_TYPELOAD](https://github.com/tpn/winsdk-10/blob/9b69fd26ac0c7d0b83d378dba01080e93349c2ed/Include/10.0.14393.0/um/CorError.h#L1484) error.

**Comment by lahma0** [comment]
> `AssemblyQualifiedName` is used when injecting .NET Core Dlls. .NET Framework Dlls just use the type name.

Ah! Yes, that would definitely explain the issue then. I will fix my code and give it another shot. Like you said, I think the documentation definitely needs clarification on this point because as far as I can tell, the documentation doesn't specifically state that .NET Framework targets only need a simple type name string (vs an assembly qualified name). 

I suppose I could have guessed based on the function signature (`string typeName` instead of `string assemblyQualifiedTypeName`) but being that the type is represented by a string, it wasn't exactly clear to me how it should be represented. 

[Off-Topic Side Note] I just noticed that the argument named `asssemblyQualifiedTypeName` in your Inject method's .NET Core overloads is misspelled (3 's' characters in 'asssembly' instead of 2) so I just figured I'd mention it in case you overlooked it.

> I think I'm going to add overload Inject methods that accept `MethodInfo` instead of string type name and method name.

That sounds like an excellent idea. I think that would make things much more clear and concise because you would then essentially have all of your assembly/namespace/type/method information all available and stored in a single MethodInfo object, making the Inject method much simpler: 
```
var methodInfo = Assembly.LoadFile(InjectDllPath)?.GetType(TypeName)?.GetMethod(MethodName);
var errorCode = targetProcess.Inject(methodInfo, string.Empty, waitForReturn);
```

> I'm actually surprised your code didn't return an error code. I ran your injector verbatim and it returned `0x80131522`, which is a [COR_E_TYPELOAD](https://github.com/tpn/winsdk-10/blob/9b69fd26ac0c7d0b83d378dba01080e93349c2ed/Include/10.0.14393.0/um/CorError.h#L1484) error.

I think the reason my code wasn't throwing an exception was because of this: 
```
if (errorCode > 0)
{
    Marshal.ThrowExceptionForHR((int)errorCode);
}
```
`0x80131522` is a negative value when represented as a signed int so `Marshal.ThrowExceptionforHR` was never being called. I should have been using `if (errorCode != 0)` instead.

Is `Marshal.ThrowExceptionForHR((int)errorCode);` the appropriate way to be translating the returned error code into a specific named exception/error? I saw the URL you referenced in your documentation which can be used to manually lookup the name of a specific error based on the error code but I wasn't sure the appropriate way to do this in code. As far as I can tell though, `Marshal.ThrowExceptionForHR` works correctly. When I run the following code: 
`Marshal.ThrowExceptionForHR(unchecked((int)0x80131522));` 
I get `TypeLoadException: Failure has occurred while loading a type.` which appears to correlate correctly to `COR_E_TYPELOAD`. I really appreciate all of your help!

**Comment by lahma0** [comment]
@Mbucari, I fixed my code and it appears that everything is working perfectly! Thanks again for all of your help and your excellent injection library. Now I just need to poke through your code and see how it actually works. Based on your documentation, I assume you're injecting shell code into the process which manually loads the .NET/.NET Framework runtime which then subsequently loads the relevant assembly? I have a lot of experience with reverse code engineering (x86/x64 assembly, IL/MSIL, IDA Pro, etc) but I don't know a whole lot about how .NET/.NET Framework work on a fundamental level. I mean, I'm well versed in IL and how the JIT compiler works and stuff but not so much in how the runtime is loaded, how it really operates, and how all the pieces fit together. I wouldn't even know where to begin to hand write assembly to do something like what you're doing. If there are any specific resources you could share that helped you in learning how to go about this, I would greatly appreciate you pointing me towards them (that is, specifically how the .NET/.NET Framework runtimes work on a fundamental level and how one would go about creating a loader). If not, no biggie. I'm sure I can learn a lot by just poking through your code. Anyways, thanks again!

**Comment by Mbucari** [comment]
> Is `Marshal.ThrowExceptionForHR((int)errorCode);` the appropriate way to be translating the returned error code into a specific named exception/error?

Yes, provided that your bootstrap method doesn't return a nonzero integer. When `Inject()` waits for return, it calls [WaitForSingleObject](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject) on the remote thread it created. The injected code returns _either_ the bootstrap method's returned int, _or_ the runtime error code if injection failed. I need to give some more thought to how this should work. I definitely want `waitForReturn` to be optional because `Bootstrap()` may be long-running, but we can't get an error code unless wait on the remote thread. It would also be nice to get the `Bootstrap()` return value separately from the runtime error code.

> I fixed my code and it appears that everything is working perfectly! Thanks again for all of your help and your excellent injection library.

Great! You're very welcome!

>Now I just need to poke through your code and see how it actually works. Based on your documentation, I assume you're injecting shell code into the process which manually loads the .NET/.NET Framework runtime which then subsequently loads the relevant assembly?

Pretty much, but I'm actually injecting raw machine code. The heart of the injector is [FrameworkLoader](https://github.com/Mbucari/InjectDotnet/blob/master/src/InjectDotnet/FrameworkLoader.cs) (.NET Framework) and [CoreLoader](https://github.com/Mbucari/InjectDotnet/blob/master/src/InjectDotnet/CoreLoader.cs) (.NET). The loader creates a struct containaing all the parameters needed by the machine code to load the framework and execute the injected dll in the target process, then writes that struct to the targets memory. It then writes the machine code into the target's memory and executes it with [ProcessExtensions.Call()](https://github.com/Mbucari/InjectDotnet/blob/15cad5125d9bd83cbce173b923e0647d7d990761/src/InjectDotnet/ProcessExtensions.cs#L141).

If you want to step through the injector assembly code, clone my repo and set a breakpoint after [Framework.Create()](https://github.com/Mbucari/InjectDotnet/blob/15cad5125d9bd83cbce173b923e0647d7d990761/src/InjectDotnet/Injector.cs#L159). Run the injector, then attach an x86/64 debugger to the target process and navigate to `loader.Loader` location. That's the machine code loader. Set a breakpoint on the loader, then resume execution of the injector.
